### PR TITLE
coprocessor: collect row count for each index prefix

### PR DESCRIPTION
**Pull request by lamxTyler**
## What have you changed? (mandatory)

In the before, we only record the row count for the whole index, this pr also records the row count for each index prefix. Therefore, for an index (a, b, c), not only could we know the row count for (a = 1 and b = 1 and c = 1), but also the row count for (a = 1) or (a = 1 and b = 1).

## What are the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

unit test and integration test with tidb.

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

PTAL @AndreMouche @coocood

### CODE REVIEW

File: src/coprocessor/codec/table.rs [hunk 1]
```diff
@@ -335,8 +335,8 @@ impl RowColsDict {
         self.cols.insert(cid, RowColMeta::new(offset, length));
     }
 
-    // get binary of cols, keep the origin order and return one slice.
-    pub fn get_column_values(&self) -> &[u8] {
+    // get binary of cols, keep the origin order, return one slice and cols' end offsets.
+    pub fn get_column_values_and_offsets(&self) -> (&[u8], Vec<usize>) {
```
Comment (coocood) [inline src/coprocessor/codec/table.rs]: We better use `end_offsets` instead of `offsets`.

File: src/coprocessor/codec/table.rs [hunk 2]
```diff
@@ -345,7 +345,12 @@ impl RowColsDict {
             }
             length += meta.length;
         }
-        &self.value[start..start + length]
+        let offsets = self.cols
```
Comment (coocood) [inline src/coprocessor/codec/table.rs]: end_offsets

File: src/coprocessor/statistics/analyze.rs [hunk 3]
```diff
@@ -87,10 +87,12 @@ impl AnalyzeContext {
             req.get_cmsketch_width() as usize,
         );
         while let Some(row) = scanner.next()? {
-            let bytes = row.data.get_column_values();
+            let (bytes, offsets) = row.data.get_column_values_and_offsets();
```
Comment (coocood) [inline src/coprocessor/statistics/analyze.rs]: ditto

**Comment by coocood** [comment]
LGTM

**Comment by breeswish** [comment]
LGTM

**Pull request merged**